## Part 1: Explore the Data

Import the data and use Pandas to learn more about the dataset.

In [1]:
import pandas as pd

In [2]:
df = pd.read_csv('Resources/client_dataset.csv')

df.head()

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year         item_id     category         subcategory  \
0          17        2023  EUD29711-63-6U        decor            wall art   
1          20        2023  XDA18116-89-4A  consumables                pens   
2           4        2023  ABE59463-05-7E     software  project management   
3          17        2023  ZMM00836-65-0C  consumables                pens   
4          21        2023  BZX55559-12-3X  consumables                misc   

   unit_price  unit_cost  unit_weight  qty  line_number  
0     1096.80     762.71         7.50  105            1  
1       24.95      15.09         1.49   21            0  
2       13.52       7.86         1.68   39            6  
3       36.42      24.85         1.23   29            3  
4      195.10     108.17        46.43   20            1

In [3]:
# View the column names in the data
df.columns
# df.columns[2] - get the column name for that index

Index(['first', 'last', 'job', 'phone', 'email', 'client_id', 'order_id',
       'order_date', 'order_week', 'order_year', 'item_id', 'category',
       'subcategory', 'unit_price', 'unit_cost', 'unit_weight', 'qty',
       'line_number'],
      dtype='object')

In [4]:
# Use the describe function to gather some basic statistics
df.describe()


client_id      order_id    order_week    order_year    unit_price  \
count  54639.000000  5.463900e+04  54639.000000  54639.000000  54639.000000   
mean   54837.869416  5.470190e+06     11.359139   2022.993064    136.267207   
std    25487.438231  2.599807e+06      7.023499      0.082997    183.873135   
min    10033.000000  1.000886e+06      1.000000   2022.000000      0.010000   
25%    33593.000000  3.196372e+06      6.000000   2023.000000     20.800000   
50%    53305.000000  5.496966e+06     11.000000   2023.000000     68.310000   
75%    78498.000000  7.733869e+06     17.000000   2023.000000    173.160000   
max    99984.000000  9.998480e+06     52.000000   2023.000000   1396.230000   

          unit_cost   unit_weight           qty   line_number  
count  54639.000000  54639.000000  5.463900e+04  54639.000000  
mean      99.446073      5.004116  5.702646e+02      2.979667  
std      133.164267      5.326599  1.879552e+04      2.436320  
min        0.010000      0.000000  0.000000e+00      0.000000  
25%       14.840000      1.450000  3.200000e+01      1.000000  
50%       49.890000      3.240000  6.800000e+01      3.000000  
75%      125.570000      6.890000  1.700000e+02      5.000000  
max      846.270000     46.430000  3.958244e+06      9.000000

In [5]:
# Use this space to do any additional research
# and familiarize yourself with the data.

df.columns[2]

'job'

In [6]:
# What three item categories had the most entries?
category_counts = df['category'].value_counts()
category_counts


category
consumables    23538
furniture      11915
software        8400
decor           7954
electronics     2832
Name: count, dtype: int64

In [7]:
# For the category with the most entries,
# which subcategory had the most entries?
subcategory_counts = df['subcategory'].value_counts()
subcategory_counts



subcategory
bathroom supplies     6424
paper products        5874
kitchen supplies      5746
tables                4540
misc                  4286
stands                4201
accounting            3756
project management    3522
plants                3466
wall art              3093
seating               2664
pens                  2603
logistics             1122
telecom                978
audio visual           732
cabling                655
desks                  510
computers              467
Name: count, dtype: int64

In [8]:
df.loc[df['category'] == 'consumables', 'subcategory'].value_counts().head(1)

subcategory
bathroom supplies    6424
Name: count, dtype: int64

In [9]:
# Which five clients had the most entries in the data?
client_entries = df['client_id'].value_counts().head()
print(client_entries)


client_id
33615    220
66037    211
46820    209
38378    207
24741    207
Name: count, dtype: int64


In [10]:
# Store the client ids of those top 5 clients in a list.
top_5_clients = list(df["client_id"].value_counts().head().index)
print(top_5_clients)


[33615, 66037, 46820, 38378, 24741]


In [11]:
df.head()

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year         item_id     category         subcategory  \
0          17        2023  EUD29711-63-6U        decor            wall art   
1          20        2023  XDA18116-89-4A  consumables                pens   
2           4        2023  ABE59463-05-7E     software  project management   
3          17        2023  ZMM00836-65-0C  consumables                pens   
4          21        2023  BZX55559-12-3X  consumables                misc   

   unit_price  unit_cost  unit_weight  qty  line_number  
0     1096.80     762.71         7.50  105            1  
1       24.95      15.09         1.49   21            0  
2       13.52       7.86         1.68   39            6  
3       36.42      24.85         1.23   29            3  
4      195.10     108.17        46.43   20            1

In [12]:
# How many total units (the qty column) did the
# client with the most entries order order?
df.loc[df['client_id'] == 33615, 'qty'].sum()

64313

## Part 2: Transform the Data
Do we know that this client spent the more money than client 66037? If not, how would we find out? Transform the data using the steps below to prepare it for analysis.

In [13]:
# Create a column that calculates the 
# subtotal for each line using the unit_price
# and the qty
df["subtotal"] = df["unit_price"] * df["qty"]
df[['unit_price', 'qty', 'subtotal']].head(2)


unit_price  qty   subtotal
0     1096.80  105  115164.00
1       24.95   21     523.95

In [14]:
# Create a column for shipping price.
# Assume a shipping price of $7 per pound
# for orders over 50 pounds and $10 per
# pound for items 50 pounds or under.

def shipping_cal (weight):
    if weight > 50:
        return weight * 7
    return weight * 10
    
df['shipping_weight'] = df['unit_weight'] * df['qty']

df['shipping_price'] = df['shipping_weight'].apply(shipping_cal)

df[['qty', 'unit_weight', 'shipping_weight', 'shipping_price']].head(2)

qty  unit_weight  shipping_weight  shipping_price
0  105         7.50           787.50          5512.5
1   21         1.49            31.29           312.9

In [15]:
# Create a column for the total price
# using the subtotal and the shipping price
# along with a sales tax of 9.25%
df['total_price'] = round((df['subtotal'] + df['shipping_price']) * 1.0925, 2)

df[['subtotal', 'shipping_price', 'total_price']].head(2)


subtotal  shipping_price  total_price
0  115164.00          5512.5    131839.08
1     523.95           312.9       914.26

In [16]:
# Create a column for the cost
# of each line using unit cost, qty, and
# shipping price (assume the shipping cost
# is exactly what is charged to the client).
df['total_cost'] = df['unit_price'] * df['qty'] + df['shipping_price']
df[['total_cost', 'qty', 'shipping_price']].head(2)


total_cost  qty  shipping_price
0   120676.50  105          5512.5
1      836.85   21           312.9

In [17]:
# Create a column for the profit of
# each line using line cost and line price
df['total_profit'] = df['total_price'] - df['total_cost']
df[['total_profit', 'total_price', 'total_cost']].head(2)


total_profit  total_price  total_cost
0      11162.58    131839.08   120676.50
1         77.41       914.26      836.85

In [18]:
df


first       last                      job               phone  \
0         Donald    Harding             Immunologist  793-904-7725x39308   
1        Tiffany      Myers          Music therapist    201.442.4543x942   
2        Shannon     Watson             Immunologist   687.737.9424x8503   
3         Nathan      Baker    Accounting technician    827-788-8123x012   
4      Christina   Schwartz             Chiropractor        265-829-3643   
...          ...        ...                      ...                 ...   
54634     Lauren      Reese           Radio producer  977.877.5272x11382   
54635    Derrick      Moore                 Musician        358.661.5483   
54636     Monica  Gutierrez         Graphic designer    294.805.9100x339   
54637      Wanda    Solomon             Toxicologist       (311)767-4924   
54638     Dwayne       Bell  Chief Financial Officer   606-619-7667x9508   

                                     email  client_id  order_id  order_date  \
0         harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1        myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                    swatson8146@payne.net      57113   1902144  2023-01-29   
3                   bakernathan@benson.com      46554   9031802  2023-04-25   
4      christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   
...                                    ...        ...       ...         ...   
54634   late_reese_4081@montoya-chavez.org      29043   9021716  2023-04-26   
54635         derrick.moore.2602@pope.info      41908   6290153  2023-02-08   
54636            gutierrezm3195@morris.org      35176   8692622  2023-03-05   
54637            solomonwanda5962@ross.org      24485   7592730  2023-01-18   
54638                     belld@galvan.com      24007   7489403  2023-01-04   

       order_week  order_year  ... unit_cost unit_weight   qty  line_number  \
0              17        2023  ...    762.71        7.50   105            1   
1              20        2023  ...     15.09        1.49    21            0   
2               4        2023  ...      7.86        1.68    39            6   
3              17        2023  ...     24.85        1.23    29            3   
4              21        2023  ...    108.17       46.43    20            1   
...           ...         ...  ...       ...         ...   ...          ...   
54634          17        2023  ...     51.60        2.25    33            8   
54635           6        2023  ...    175.46       11.70    47            0   
54636           9        2023  ...     57.31        4.16   475            6   
54637           3        2023  ...      1.22       18.04   112            0   
54638           1        2023  ...      1.61        2.07  1031            5   

        subtotal  shipping_weight  shipping_price  total_price  total_cost  \
0      115164.00           787.50         5512.50    131839.08   120676.50   
1         523.95            31.29          312.90       914.26      836.85   
2         527.28            65.52          458.64      1077.12      985.92   
3        1056.18            35.67          356.70      1543.57     1412.88   
4        3902.00           928.60         6500.20     11364.40    10402.20   
...          ...              ...             ...          ...         ...   
54634    2743.29            74.25          519.75      3564.87     3263.04   
54635    9709.73           549.90         3849.30     14813.24    13559.03   
54636   31188.50          1976.00        13832.00     49184.90    45020.50   
54637     165.76          2020.48        14143.36     15632.71    14309.12   
54638    3103.31          2134.17        14939.19     19711.43    18042.50   

       total_profit  
0          11162.58  
1             77.41  
2             91.20  
3            130.69  
4            962.20  
...             ...  
54634        301.83  
54635       1254.21  
54636       4164.40  
54637       1323.59  
54638       1668.93  

## Part 3: Confirm your work
You have email receipts showing that the total prices for 3 orders. Confirm that your calculations match the receipts. Remember, each order has multiple lines.

Order ID 2742071 had a total price of \$152,811.89

Order ID 2173913 had a total price of \$162,388.71

Order ID 6128929 had a total price of \$923,441.25


In [19]:
# Check your work using the totals above - first make a list with the order numbers, run through the order list 
order_ids = [2742071, 2173913, 6128929]
#order_ids is a list containing some order IDs.

def order_price (order_id):
    order_lines_df = df.loc[df['order_id'] == order_id, "total_price"]
    return round(order_lines_df.sum(), 2)
#There's a function defined named find_order_price(order_id). This function takes an order ID as input and does the following:
#It locates the rows in a DataFrame df where the 'order_id' column matches the given order ID.
#Then, it selects the values from the 'line_price' column corresponding to those rows.
#Finally, it sums up those values and rounds the result to two decimal places before returning it.

for order_id in order_ids:
    print(f"{order_id}, ${order_price(order_id)}")
#A loop iterates over each order ID in the order_ids list. For each order ID, it prints out a message
#showing the order ID and its total price by calling the find_order_price() function for that order ID.

2742071, $152811.89
2173913, $162388.71
6128929, $923441.25


## Part 4: Summarize and Analyze
Use the new columns with confirmed values to find the following information.

In [22]:
import pandas as pd

In [27]:
# How much did each of the top 5 clients by quantity
# spend? Check your work from Part 1 for client ids.
#def qty_spend 
def client_totals(client_id, column):
    client_df = df.loc[df['client_id'] == client_id, column]
    return round(client_df.sum(), 2)
for client_id in top_5_clients:
    print(f"{client_id}: ${client_totals(client_id, 'total_price')}")

33615: $8377308.52
66037: $10259514.79
46820: $9743794.36
38378: $12906550.87
24741: $82268892.02


In [32]:
# Create a summary DataFrame showing the totals for the
# for the top 5 clients with the following information:
# total units purchased, total shipping price,
# total revenue, and total profit. Sort by total profit.

data_dict = {'client_id': top_5_clients}
columns = ['qty', 'shipping_price', 'total_price', 'total_cost', 'total_profit']
for col in columns:
    data_dict[col] = [client_totals(client_id, col) for client_id in top_5_clients]
    
summary_df = pd.DataFrame(data_dict)
summary_df

client_id     qty  shipping_price  total_price   total_cost  total_profit
0      33615   64313      1828984.89   8377308.52   7668017.00     709291.52
1      66037   43018      1395151.85  10259514.79   9390860.23     868654.56
2      46820   75768      1601448.84   9743794.36   8918804.87     824989.49
3      38378   73667      3429455.40  12906550.87  11813776.55    1092774.32
4      24741  239862      5126448.37  82268892.02  75303333.62    6965558.40

In [34]:
# Format the data and rename the columns
# to names suitable for presentation.
# Currency should be in millions of dollars.

# make sure the variables are suited for you data.

money_cols = ['total_price', 'shipping_price', 'total_cost', 'total_profit']

def currency_format_millions(amount):
    return amount/1000000

for col in money_cols:
    summary_df[col] = summary_df[col].apply(currency_format_millions)

rename_dict = {
    'client_id': 'Client ID',
    'line_price': 'Total Revenue (millions)',
    'shipping_price': 'Shipping (millions)',
    'qty': 'Units',
    'line_cost': 'Total Cost (millions)',
    'line_profit': 'Total Profit (millions)'
}
summary_df = summary_df.rename(columns = rename_dict)
summary_df

Client ID   Units  Shipping (millions)  total_price  total_cost  \
0      33615   64313             1.828985     8.377309    7.668017   
1      66037   43018             1.395152    10.259515    9.390860   
2      46820   75768             1.601449     9.743794    8.918805   
3      38378   73667             3.429455    12.906551   11.813777   
4      24741  239862             5.126448    82.268892   75.303334   

   total_profit  
0      0.709292  
1      0.868655  
2      0.824989  
3      1.092774  
4      6.965558

In [38]:
# Sort the updated data by "Total Profit" form highest to lowest

summary_df = summary_df.sort_values('total_profit', ascending=False)
summary_df

Client ID   Units  Shipping (millions)  total_price  total_cost  \
4      24741  239862             5.126448    82.268892   75.303334   
3      38378   73667             3.429455    12.906551   11.813777   
1      66037   43018             1.395152    10.259515    9.390860   
2      46820   75768             1.601449     9.743794    8.918805   
0      33615   64313             1.828985     8.377309    7.668017   

   total_profit  
4      6.965558  
3      1.092774  
1      0.868655  
2      0.824989  
0      0.709292